docs/architecture/adr-001-coin-source-tracing.md

```diff
@@ -0,0 +1,260 @@
+    trace = portID + "/" + channelID +"/" + trace + "/"
+  }
+  
+  traceHash = tmbytes.HexBytes(tmhash.Sum(trace))
+
+  // set the value to the lookup table if not stored already
+  if !k.HasTrace(ctx, traceHash) {
+    k.SetTrace(ctx, traceHash)
+  }
+
+  denom = "ibc/"+ traceHash.String()  "/" + baseDenom
+  return denom
+}
+```
+
+The denomination also needs to be updated when token is received on the source chain:
+
+```golang
+// UnprefixDenom removes the first portID/channelID pair from a given denomination trace info and returns the
+// denomination with the updated trace hash and the new trace info.
+// An error is returned if the trace cannot be found on the store from the denom's trace hash.
+func (k Keeper) UnprefixDenom(ctx Context, denom string) (denom, trace string, err error) {
+  denomSplit := strings.SplitN(denom, "/", 3)
+
+  switch {
+    case denomSplit[0] == denom:
+      return "", Wrapf(ErrInvalidDenomForTransfer, "denomination should be prefixed with the format 'ibc/{traceHash}/%s'", denom)
+    case denomSplit[0] != "ibc":
+      return "", Wrapf(ErrInvalidDenomForTransfer, "denomination %s must start with 'ibc'", denom)
+  }
+
+  baseDenom = denomSplit[2]
+  traceHash := tmbytes.HexBytes(denomSplit[1])
+  // Get the value from the map trace hash -> trace info prefix
+  trace = k.GetTrace(ctx, traceHash)
+  if trace == "" {
+    return "", Wrapf(ErrTraceNotFound, "denom: %s", denom)
+  }
+
+  traceSplit := strings.SplitN(denom, "/", 3)
+  if len(traceSplit) == 3 {
+    // the trace has only one portID/channelID pair
+    return baseDenom
+  }
+
+  // remove a single identifiers pair to create the new trace
+  trace = trace[2]
+  traceHash = tmbytes.HexBytes(tmhash.Sum(trace))
+
+  // set the value to the lookup table if not stored already
+  if !k.HasTrace(ctx, traceHash) {
+    k.SetTrace(ctx, traceHash)
+  }
+
+  denom = "ibc/"+ traceHash.String() + "/" + baseDenom
+  return denom
+}
+```
+
+```golang
+// GetTraceFromDenom returns the token source tracing info from the trace hash of the given denomination
+func (k Keeper) GetTraceFromDenom(ctx Context, denom string) (string, error) {
+  denomSplit := strings.Split(denom, "/")
+  if denomSplit[0] == denom {
+    return "", nil
+  }
+
+  traceHash := tmbytes.HexBytes(denomSplit[1])
+  // Get the value from the map trace hash -> trace info prefix
+  trace := k.GetTrace(ctx, traceHash)
+  if trace == "" {
+    return "", Wrapf(ErrTraceNotFound, "denom: %s", denom)
+  }
+
+  return trace
+}
+
+```
+
+Additionally, the `SendTransfer`'s `createOutgoingPacket` call and the `OnRecvPacket` need to be
+updated to be retreive the trace info (using `GetTraceFromDenom`) prior to checking the correctness
+of the prefix.
+
+### Coin Changes
+
+The coin denomination validation will need to be updated to reflect these changes. In particular, the denomination validation
+function will now accept slash separators (`"/"`) and will bump the maximum character length to 64.
```

Comment:
I like the idea of using `traceHash` instead of storing the full `Trace` in each coin. However, the `TraceHash` still cannot be part of the denomination.

The reason for this is because the length validation of the `baseDenom` is getting mixed in with the length validation of the total `denom+tracehash`.

Suppose, we have a coin with a baseDenom of 64 characters. This will still pass the original length validation, but it will fail as soon as we add a tracehash.

I recommend we keep denomination to only keeping the baseDenom, and have the `traceHash` as a separate field in the `Coin` struct

```
Denom       string // length-validated
TraceHash [256]byte
```

This is a small change w.r.t. where the traceHash is retrieved from, but everything else stays the same.

Whether or not users are presented with `ibc/{traceHash}/baseDenom` or anything else is a UI concern.

Reply:
Looking at discussion here: https://github.com/cosmos/cosmos-sdk/pull/6662#discussion_r452266404

I agree with keeping the traceHash in the denom, though I do think that we still want cleaner length validation logic.

Given the above discussion I would propose:

```
denomSplit := strings.Split(denom, "/")
if denomSplit[len(denomSplit)-1] > maxBaseLength {
     return err
}
if denom > maxTotalLength {
    return err
}
```

This removes the potential issue I describe here: https://github.com/cosmos/cosmos-sdk/pull/6662#discussion_r459613291